### ISSUE TITLE
14.1-Introduction of Deep Reinforcement Learning (RL)

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-HYL#73]
Deep Reinforcement Learning = RL 增強式學習

到目前為止啊 我們講的幾乎都是 Supervised Learning
在講 Self Supervised Learning 的時候 我們其實也是很類似 Supervised Learning 的方法
只是我們的 Label 不需要特別僱用人力去標記 它可以自動產生

在講 Auto-encoder 的時候雖然說它是一個 Unsupervised pre-train的方法
我們沒有用到人類的標記 但事實上 我們還是有一個 Label
只是這個 Label 不需要耗費人類的力量來產生而已

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/1d4d37e6-c615-4f14-a3cf-7fb08dd410e0)

假設你要叫機器學習下圍棋
用 Supervised Learning 的方法 就是告訴機器說
看到現在的盤勢長這個樣子的時候 下一步應該落子的位置在哪裡
但是問題是哪一個是最好的下一步?

```
使用Deep Reinforcement Learning增強式學習的時機

不知道最佳的輸出應該是什麼(正確答案人類也不知道是什麼的時候 )
要蒐集有標註的資料很困難的時候
```

Machine Learning 就是三個步驟 
RL 也是一模一樣的三個步驟

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/1d78d77a-c287-4979-8d28-b7f899d430e6)

Reinforcement Learning 裡面會有一個 Actor 還有一個 Environment 
Environment 環境會給 Actor 一個 Observation

Actor 跟 Environment會進行互動 
在這個互動的過程中呢這個 Environment 會不斷地給這個 Actor 一些 Reward 
告訴它說你現在採取的這個 Action 它是好的還是不好的
今天要找的這個 Function 可以拿 Observation 當作 Input, Action 當作 Output 
Function 的目標 是要去 Maximizing 我們可以從 Environment 獲得到的 Reward 的總和

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/95d80dbd-0a56-4880-8373-8050251d05e4)
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/509660e0-5388-4329-b1e1-5a66923e3733)
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/b8bdd2e8-dd8d-41a6-942d-3f09f13d0df2)

如果今天你要讓機器來下圍棋 那你的 Actor 就是就是 AlphaGo
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/4893d98e-22f4-40ec-80a8-671fb65d8c40)
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/d7481e85-f659-4f13-aea4-35f3aea1674d)

你會定義說 如果贏了就得到 1 分 如果輸了就得到 -1 分
也就是說在下圍棋這整個這個你的 Actor 跟環境互動的過程中
只有遊戲結束你才能夠拿到 Reward 就你最後
在中間整個互動的過程中的 Reward就都算是 0 分 沒有任何的 Reward

________________________

### RL 跟機器學習的 Framework

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/964d06b7-bbd6-42ae-9f06-899abbdb687c)

RL 其實也是一模一樣的三個步驟

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/2f69f1e3-62d9-4ad9-bde9-4ad5a9b15f90)

第一步 有未知數的 Function 就是我們的 Actor
在 RL 裡面 你的 Actor 就是一個 Network  = Policy 的 Network

玩 Space Invader 當作例子的話
這個遊戲畫面上的 Pixel 像素就是這一個 Actor 的輸入
輸出就是 每一個可以採取的行為
它的分數 每一個可以採取的 Action 它的分數
舉例來說  向左 0.7 分 向右 0.2 分 開火 0.1 分 

那最後機器會決定採取哪一個 Action 取決於每一個 Action 取得的分數

常見的做法是直接把這個分數 就當做一個機率
然後按照這個機率去 Sample 去隨機決定要採取哪一個 Action
舉例來說 向左得到 0.7 分 那就是有 70% 的機率會採取向左
採取 Sample 有一個好處是說 今天就算是看到同樣的遊戲畫面 你的機器每一次採取的行為也會略有不同 
在很多的遊戲裡面這種隨機性是重要的
比如說你在做剪刀石頭布的時候 
如果你總是會出石頭 就跟小叮噹一樣 那你就很容易被打爆

第二步 RL 定義 Loss
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/a7d1b3bd-316f-442e-a812-81f5f3b54d44)

輸入的遊戲畫面 我們叫它 s1 
輸出的 Action 就叫它 a1

接下來機器會看到新的遊戲畫面又會採取新的行為
反覆持續下去直到遊戲結束了
從遊戲開始到結束的這整個過程稱之為一個 Episode
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/c5b0aae7-e4f5-417f-b5af-5f268f291b6d)

每一個行為都可能得到 Reward (r)
把所有的 Reward 通通集合起來  = 整場遊戲的 Total Reward = R = Return

Loss 是越小越好 -> 負的 Total Reward 就當做我們的 Loss

第三步 RL Optimization
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/2bc08487-7a2e-474e-8c8e-f12048f9f7f5)

s 跟 a 所形成的 Sequence 就是 s1 a1 s2 a2 s3 a3 這個 Sequence = 又叫做 Trajectory = τ 
Trajectory = （射體在空中的）軌道，軌跡

根據這個互動的過程 Machine 會得到 Reward
決定 Reward 的時候 光看 Action 是不夠的 還要看現在的 Observation 才可以
因為並不是每一次開火你都一定會得到分數 
開火要正好有擊到外星人 外星人正好在你前面你開火才有分數
所以 Reward 是一個 Function ->它拿 a1 跟 s1 當作輸入 輸出r1 
它拿 a2 跟 s2 當作輸入產生 r2 作為輸出 

全部結合起來就得到 R =  Total Reward = r1+r2+r3+... ->我們要去 Maximize 的對象

```
Optimization 
你要去 Learn 出一組Network的參數
這一組參數放在 Actor 的裡面它可以讓這個 R 的數值越大越好
```

但是 RL 困難的地方是這不是一個一般的 Optimization 的問題
因為你的 Environment 這邊有很多問題導致說它跟一般的 Network Training 不太一樣 

第一個問題是 你的 Actor 的輸出是有隨機性的

action它是用 Sample 產生的 你定同樣的 s1 每次產生的 a1 不一定會一樣

所以假設你把 Environment Actor 跟 Reward 合起來當做是一個巨大的 Network 來看待
這個 Network 可不是一般的 Network 這個 Network 裡面是有隨機性的
這個 Network 裡面某一個 Layer 是 它的輸出每次都是不一樣的

另外還有一個更大的問題就是 Environment 跟 Reward 它根本就不是 Network
它只是一個黑盒子而已 你根本不知道裡面發生了什麼事情 
Environment 就是遊戲機 裡面發生什麼事情你不知道 
你只知道說你輸入一個東西會輸出一個東西
你採取一個行為它會有對應的回應

Reward 可能比較明確 但它也不是一個 Network 它是一條規則
看到這樣子的 Optimization 跟這樣的 Action會得到多少的分數 

更麻煩的地方是 往往 Reward 跟 Environment它也是有隨機性的

如果是在電玩裡面 通常 Reward 可能比較不會有隨機性(因為規則是定好的)

但是在 Environment 裡面就算是在電玩的這個應用中它也是有隨機性的
如果是下圍棋 你落子在同一個位置
對手會怎麼樣回應每次可能也是不一樣
所以環境很有可能也是有隨機性的

RL 真正的難點就是我們怎麼解這一個 Optimization 的問題
怎麼找到一組 Network 參數可以讓 R 越大越好

這整個問題跟 GAN 其實有異曲同工之妙
它們有一樣的地方 也有不一樣的地方

先說它們一樣的地方在哪裡
在訓練 Generator 的時候 你會把 Generator 跟 Discriminator 接在一起
然後你希望去調整 Generator 的參數 讓 Discriminator 被騙過(輸出越大越好?)
在 RL 裡面 我們也可以說這個 Actor 就像是 Generator
Environment 跟 Reward 就像是 Discriminator

不一樣的地方
在 GAN 裡面你的 Discriminator 也是一個 Neural Network
你了解 Discriminator 裡面的每一件事情 它也是一個 Network 
你可以用 Gradient Descent來 train 你的 Generator 讓 Discriminator 得到最大的輸出
但是在 RL 的問題裡面你的 Reward 跟 Environment 你可以把它們當 Discriminator 來看
但它們不是 Network(它們是一個黑盒子) ->所以你沒有辦法用 一般 Gradient Descent 的方法來調整你的參數
